lahja/trio/endpoint.py

```diff
@@ -0,0 +1,905 @@
+        # (ConnectionConfig, trio.Event) which is retrieved by
+        # `_process_connections`.
+        (self._connection_send_channel, connection_receive_channel) = cast(
+            ConnectionChannelPair, trio.open_memory_channel(100)
+        )
+
+        # This send/receive channel is fed by `Endpoint.broadcast_nowait`
+        # which places a 2-tuple of (BaseEvent, BroadcastConfig) which is
+        # retrieved by the `_process_broadcasts` and fed into the
+        # `Endpoint.broadcast` API.
+        (self._broadcast_send_channel, broadcast_receive_channel) = cast(
+            ConcreteBroadcastChannelPair, trio.open_memory_channel(100)
+        )
+
+        self.logger.debug("%s: starting", self)
+
+        async with trio.open_nursery() as nursery:
+            #
+            # _process_broadcasts:
+            #     Manages a channel that `Endpoint.broadcast_nowait` places
+            #     `Broadcast` messages on (typically from synchronous
+            #     contexts).  These messages are fed into
+            #     `Endpoint.broadcast`
+            #
+            nursery.start_soon(
+                self._process_broadcasts, broadcast_receive_channel, nursery
+            )
+
+            #
+            # _process_received_messages:
+            #     Manages a channel which all incoming
+            #     event broadcasts are placed on, running each event through
+            #     the internal `_process_item` handler which handles all of the
+            #     internal logic for the various
+            #     `request/response/subscribe/stream/wait_for` API.
+            #
+            nursery.start_soon(
+                self._process_received_messages, message_receive_channel, nursery
+            )
+
+            #
+            # _process_connections:
+            #     When `Endpoint.connect_to_endpoint` is called, the actual
+            #     connection process is done asynchronously by putting the
+            #     `ConnectionConfig` onto a queue which this process
+            #     retrieves to establish the new connection.  This includes
+            #     handing the `RemoteEndpoint` oblect off to the `RemoteManager`
+            #     which takes care of the connection lifecycle.
+            #
+            nursery.start_soon(
+                self._process_connections, connection_receive_channel, nursery
+            )
+
+            #
+            # _monitor_subscription_changes
+            #    Monitors an event for local changes to subscriptions to
+            #    propagate those changes to remotes.
+            nursery.start_soon(self._monitor_subscription_changes)
+
+            # wait for the sub-processes to start before marking the endpoint
+            # as running.
+            await self._message_processing_loop_running.wait()
+            await self._connection_loop_running.wait()
+            await self._process_broadcasts_running.wait()
```

Comment:
I think all these could be removed with `Nursery.start`

Reply:
Actually, looking at it again, if they're waiting on channels there's no need to wait for them to start at all! The channels will queue up any interactions with them which occur before they start.